src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,329 @@
+
+  // Get adjusted fitness.
+  double AdjustedFitness() const { return aAdjustedFitness; }
+
+  // Get neuron number.
+  ssize_t NumNeuron() const {
+    return aNeuronGenes.size();
+  }
+  
+  // Get link number.
+  ssize_t NumLink() const {
+    return aLinkGenes.size();
+  }
+
+  // Whether specified neuron id exist in this genome.
+  bool HasNeuronId(ssize_t id) const {
+    for (ssize_t i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Id() == id) {
+        return true;
+      }
+    }
+    
+    return false;
+  }
+
+  // Get neuron by id.
+  NeuronGene GetNeuronById(ssize_t id) {
+    assert(HasNeuronId(id));
+
+    for (ssize_t i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Id() == id) {
+        return aNeuronGenes[i];
+      }
+    }
+  }
+
+  // Get neuron index by id.
+  ssize_t GetNeuronIndex(ssize_t id) const {
+    for(ssize_t i=0; i < NumNeuron(); ++i) {
+        if (aNeuronGenes[i].Id() == id) {
+            return i;
+        }
+    }
+
+    return -1;  // Id start from 0.
+  }
+
+  // Get link index by innovation id.
+  ssize_t GetLinkIndex(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId) {
+            return i;
+        }
+    }
+
+    return -1;  // Id start from 0.
+  }
+
+  // Whether link exist and enabled.
+  bool ContainEnabledLink(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId &&
+            aLinkGenes[i].Enabled()) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Whether link exist.
+  bool ContainLink(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Set neurons' input and output to zero.
+  void Flush() {
+    for (ssize_t i=0; i<aNeuronGenes.size(); ++i) {
+      aNeuronGenes[i].Activation(0);
+      aNeuronGenes[i].Input(0);
+    }
+  }
+
+  // Sort link genes by toNeuron's depth.
+  void SortLinkGenes() {
+    struct DepthAndLink
+    {
+      double depth;
+      LinkGene link;
+ 
+      DepthAndLink(double d, LinkGene& l) : depth(d), link(l) {}
+
+      bool operator < (const DepthAndLink& dL) const
+      {
+        return (depth < dL.depth);
+      }
+    };
+
+    std::vector<double> toNeuronDepths;
+    for (ssize_t i=0; i<aLinkGenes.size(); ++i) {
+      NeuronGene toNeuron = GetNeuronById(aLinkGenes[i].ToNeuronId());
+      toNeuronDepths.push_back(toNeuron.Depth());
+    }
+
+    std::vector<DepthAndLink> depthAndLinks;
+    ssize_t linkGenesSize = aLinkGenes.size();
+    for (ssize_t i=0; i<linkGenesSize; ++i) {
+      depthAndLinks.push_back(DepthAndLink(toNeuronDepths[i], aLinkGenes[i]));
+    }
+
+    std::sort(depthAndLinks.begin(), depthAndLinks.end());
+
+    for (ssize_t i=0; i<linkGenesSize; ++i) {
+      aLinkGenes[i] = depthAndLinks[i].link;
```

Comment:
I think, we could speed this up, if we would just store the index of all linkGenes, an use another list, that contains all linkGenes as a reference.